File: examples/typescript/boundary-parsing.ts
// Parse, don't validate: branded types and type guard parsers.

// --- Branded types ---

type Email = string & { readonly __brand: "Email" };
type Age = number & { readonly __brand: "Age" };

interface ParseError {
  readonly field: string;
  readonly message: string;
}

// --- Parsing (returns typed value or error) ---

const EMAIL_RE = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;

function parseEmail(raw: unknown): Email | ParseError {
  if (typeof raw !== "string") {
    return { field: "email", message: "must be a string" };
  }
  const trimmed = raw.trim();
  if (!EMAIL_RE.test(trimmed)) {
    return { field: "email", message: `invalid email format: ${JSON.stringify(trimmed)}` };
  }
  return trimmed as Email;
}

function parseAge(raw: unknown): Age | ParseError {
  if (typeof raw !== "number" || !Number.isInteger(raw)) {
    return { field: "age", message: "must be an integer" };
  }
  if (raw < 0 || raw > 150) {
    return { field: "age", message: `must be 0-150, got ${raw}` };
  }
  return raw as Age;
}

function isParseError(v: unknown): v is ParseError {
  return typeof v === "object" && v !== null && "field" in v && "message" in v;
}

// --- ANTI-PATTERN: Validate, then use raw ---

function isValidEmail(raw: string): boolean {
  return EMAIL_RE.test(raw.trim()); // caller still has untyped string
}

function sendWelcomeBad(raw: string): void {
  if (isValidEmail(raw)) {
    console.log(`Sending to ${raw}`); // raw could have spaces, etc.
  }
}

// --- CORRECT: Use parsed value ---

function sendWelcome(email: Email): void {
  console.log(`Sending to ${email}`);
}

export { Email, Age, ParseError, parseEmail, parseAge, isParseError, sendWelcome };

File: examples/python/boundary_parsing.py
"""Parse, don't validate: convert raw input to typed domain values or structured errors."""
from dataclasses import dataclass
from typing import Tuple
import re


@dataclass(frozen=True)
class Email:
    local: str
    domain: str

    def __str__(self) -> str:
        return f"{self.local}@{self.domain}"

@dataclass
class ParseError:
    field: str
    message: str


# --- CORRECT: Parse into a typed value ---

_EMAIL_RE = re.compile(r"^([a-zA-Z0-9._%+-]+)@([a-zA-Z0-9.-]+\.[a-zA-Z]{2,})$")

def parse_email(raw: str) -> Tuple[Email | None, ParseError | None]:
    raw = raw.strip()
    match = _EMAIL_RE.match(raw)
    if not match:
        return None, ParseError("email", f"invalid email format: {raw!r}")
    return Email(local=match.group(1), domain=match.group(2)), None


@dataclass(frozen=True)
class Age:
    value: int

def parse_age(raw: object) -> Tuple[Age | None, ParseError | None]:
    if not isinstance(raw, int) or raw < 0 or raw > 150:
        return None, ParseError("age", f"must be an integer 0-150, got {raw!r}")
    return Age(value=raw), None


# --- ANTI-PATTERN: Validate then use raw data ---

def is_valid_email(raw: str) -> bool:
    """Validation returns bool — caller still has an untyped string."""
    return bool(_EMAIL_RE.match(raw.strip()))

def send_welcome(raw_email: str) -> None:
    """Bug-prone: uses raw string after a bool check somewhere upstream."""
    if is_valid_email(raw_email):
        print(f"Sending to {raw_email}")  # raw_email could have spaces, etc.


# --- CORRECT: Use parsed value ---

def send_welcome_safe(email: Email) -> None:
    """Type-safe: only accepts already-parsed Email."""
    print(f"Sending to {email}")


if __name__ == "__main__":
    email, err = parse_email("  user@example.com  ")
    if email:
        send_welcome_safe(email)
    age, err = parse_age(25)
    if age:
        print(f"Age: {age.value}")
